server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -56,11 +57,31 @@ public class DesiredBalanceShardsAllocator implements ShardsAllocator {
 
     private final ShardsAllocator delegateAllocator;
     private final ThreadPool threadPool;
+    /**
+     * This is a callback to run {@link AllocationService#executeWithRoutingAllocation(ClusterState, String, RerouteStrategy)}, which
+     * produces a new ClusterState with the changes made by {@link DesiredBalanceReconciler#reconcile}. The {@link RerouteStrategy} provided
+     * to the callback calls into {@link #desiredBalanceReconciler} for the changes. The {@link #masterServiceTaskQueue} will apply the
+     * cluster state update.
+     */
     private final DesiredBalanceReconcilerAction reconciler;
     private final DesiredBalanceComputer desiredBalanceComputer;
+    /**
+     * Reconciliation ({@link DesiredBalanceReconciler#reconcile(DesiredBalance, RoutingAllocation)}) takes the {@link DesiredBalance}
+     * output of {@link DesiredBalanceComputer#compute} and identifies how shards need to be added, moved or removed to go from the current
+     * cluster shard allocation to the new desired allocation.
+     */
     private final DesiredBalanceReconciler desiredBalanceReconciler;
```

Comment:
Should this doc be on `DesiredBalanceReconciler#reconcile` ?

Reply:
I've updated the PR with a new comment on `DesiredBalanceReconciler#reconcile`

I'd like to explain here how `desiredBalanceReconciler` differs from `reconciler`. They have practically the same name right now, so I think it makes the code more understandable to very clearly explain how they are used / what they do in this file's context.